### DISCUSSION TITLE
Access Token is unusable to authenticate on backend

### DISCUSSION THREAD

**Discussion by MathBeaudoin** [supabase/supabase#31034] [Questions]
Hello! To start this thread, here is my use case: 

- I have a web application (Vue 3) that authenticates users with OAuth
- I want to use send the `access_token` generated to my backend to accept authenticated requests

BUT 

- When sending the access_token as `Authorization: Bearer access_token_value_here,` a 401 (Unauthorized) is always returned
- After further investigation, on jwt.io, the token sent in `Authorization: Bearer ...` has an invalid signature

I am getting my access token by using the session of a Supabase instance:
```
    public async getAuthToken{
        return await this.supabase.auth.getSession().data.session.access_token;
    }
```

On jwt.io, the header and payload of my token have the right values. It's the signature that can't be validated. I assume I am getting the 401 Unauthorized because `Quarkus` is not letting me use JWT with invalid signatures.

Why do my tokens have an invalid signature? What would be the best way for me to authenticate to my own backend with Supabase?

**Comment by GaryAustin1** [comment]
Are you using the JWT secret from your instance to validate the JWT signature. ?

**Reply by MathBeaudoin** [reply]
Since I do not want to store the JWT secret on the SPA: no, the JWT signature is not yet validated when sending an HTTP request from the SPA to the server with `Authorization: Bearer access_token_here`. 

I would have ideally wanted to validate the signature on the backend.

**Reply by GaryAustin1** [reply]
What is generating the 401?   If you are calling Supabase with it you also need the anon key set in the apikey header.

**Reply by GaryAustin1** [reply]
The JWT secret can only be used server side if you are trying to validate the JWT yourself for use on your own server code.    Or you can use getUser.

**Reply by MathBeaudoin** [reply]
After hours of research, it seems like the problem was server side. I am sorry for the inconveniance, tyvm.

**Reply by cravenceiling** [reply]
> After hours of research, it seems like the problem was server side. I am sorry for the inconveniance, tyvm.

How did you solve the issue? Are you still using getSession()?

I'm having a similar problem, I want to use the token stored in the cookie (sb-\<project-id\>-auth-token) instead of using getSession(), but I get the token encoded (base64) so I can't validate it in my backend. I'm using the supabase secret to verify the token.

**Reply by MathBeaudoin** [reply]
@timetravel-1010 The problem was entirely on the backend with Quarkus. Here are some notes I wrote myself to remember the procedure in case it happened to me again:

## Authenticate user with JWT: A step by step guide
- Resource: https://github.com/quarkusio/quarkus/discussions/42550

1. Create file `secret_supabase_key.jwk` with content :
```
{
    "kty": "oct",
    "k": "",
    "alg": "HS256"
}
```

2. Find Supabase JWT Secret in dashboard, and convert it in https://base64.guru/converter/encode/text

3. Add the converted result to `k` field


Now, there is two ways to enable the JWT authentication:

4.1.1. Add `smallrye.jwt.verify.key.location=secret_supabase_key.jwk` to `application.properties`

OR

4.2.1. Take the whole file's content, and convert it in https://base64.guru/converter/encode/text

4.2.2. Add `smallrye.jwt.verify.secretkey=result_of_the_converted_file` (this value should be in a .env and loaded dynamically)

4.2.3. Delete file `secret_supabase_key.jwk`

JWT authentication should be good to go! The second option is probably the ideal one to use, since a single .env value could be stored.
